### ISSUE TITLE
Submodule Support

### ISSUE THREAD

**Issue by rcdailey** [petervanderdoes/gitflow-avh#216]
I feel like I'm touching on a subject that has been discussed before, however I haven't seen anything in the list of closed issues about it.

Is it possible to automate submodules via gitflow scripts? For example, when branching for a release, the script should also branch the submodule. Likewise when hotfixing or merging in release branches into the master branch.

Or is it just better to repeat git flow commands by hand in each submodule, going from bottom up?

**Comment by petervanderdoes** [comment]
To be honest, I don't know. I try to avoid submodules and have succeeded so far.
I'll let others chime in on that one.

**Comment by rcdailey** [comment]
Avoiding submodules becomes difficult when you want to accomplish both of the below tasks.

1. Have a common code base shared between multiple projects or applications
2. Keep your repository as small as possible and have separation of concerns

Git has no sparse checkouts and while yes adding submodules does add workflow complexity, the benefits outweigh the productivity hit. Regardless of our personal stances on submodules though, I think submodules are perfectly functional in the gitflow model and if there is room for automation I think we should explore it.

Thanks for taking the time to hear me out.

**Comment by petervanderdoes** [comment]
I don't think you should ever work on a submodule from within another application. If the submodule needs to be changed you need to work on this repository by itself and then update the submodule in your application repository. Changing a submodule has a bigger impact than just the application you are working on.

So the question is, if you update the submodule,  `git submodule update`, in a feature branch does the update get merged back in develop?

**Comment by rcdailey** [comment]
This is a bigger issue than simply updating the submodule. When I hotfix a release, and if by chance changes to the submodule are required for the hotfix, I want a hotfix branch to be created in the submodule repository, based on the commit tied to the hotfix branch in the parent repository. This will allow modifications to the correct revision in the submodule without interruptions from other mainline branches.

The principles are all the same for submodules. If you think of it as a physical piece of the repository, it makes more sense.

I realize that this doesn't apply to *all* submodules, but should apply to those which it makes sense (i.e. make it configurable).

**Comment by petervanderdoes** [comment]
git-flow is based on branches, not SHA, so if you want to git-flow a submodule you have to do a branch, which will work for submodules. Now you have changed the branch, you push it to a shared repository, I'm assuming the submodule is located somewhere shareable. When somebody else, in another project, runs a git submodule update, it will pull in your changes and that can potentially be a problem.

Technically I think this might be possible. Looking at it from an ITIL Change Management view I just don't see it. A change that could potentially crash other applications can never been done in a change for one particular application.

Maybe I'm just not seeing your use of submodules. I always think of shareable code between multiple projects, and than I rather use a dependency management system, like composer, rubygems, maven.

**Comment by allancarter** [comment]
We are using submodules too and I'd like git flow to handle submodules.

Regarding whether git submodule update in other projects would get your changes, the answer is no. Not unless they go into the submodule and pull the latest version. If they are working on the develop branch then the super-project and all submodules should be on the same branch: develop.

It would be nice to have a --recurse-submodules switch similar to git that would branch the super-project and all submodules recursively when a new branch (feature, hotfix, release, etc.) is created and finish them all together too.

Right now I'm using "git submodule foreach" to do this which simplifies things somewhat. However, I'm hitting a bug where "git flow feature finish" errors out when finishing a branch that doesn't have any changes to merge back to develop. I'll report a separate issue for that.

**Comment by charles4711** [comment]
I'd really like to see the submodule thing also. IMHO submodules are such a good way to handle dependencies on a low level basis without the need to use a tool for that. In some projects we didn't use external tools for handling dependencies. Submodules have just too many pitfalls to use them correctly, so maybe adding support for submodules in git-flow would ease up.
In one of my current projects I'm writing Matlab applications using Libraries, which are self-developed. Instead of using each submodule for itself and integrating later, it's just faster using submodules.

I usually prefer the following steps:
1. Create a new feature in the top module
2. Create the same feature branch in all sub modules (usually just 1 level)
3. edit the .gitmodules file in the top module to point to the submodule feature branch
4. do a commit in both sub and top module
5. call git submodule update --remote to point to the correct submodule branch
6. Develop ...

Merging:
7. Merge submodules first bottom up.
8. Merge top repo.
9. Edit .gitmodules to point to correct branch

It is really important to branch the submodules too, because we always end up in a detached head state in the submodules once submodule update was called...

So there is some sort of automation possible
